### ISSUE TITLE
Feature Request: Length Masking for RNNs

### ISSUE THREAD

**Issue by shawnjhenry** [pytorch/pytorch#517]
Some models with RNN components require batching different length inputs by zero padding them to the same length.  The zero padding can interfere with proper training of the network since RNNs with non-zero biases in the linear layers can have non-zero output even with zero input and state.  

One solution to this problem is to pass in an optional parameter of size batch_size containing the lengths of the sequences, allowing the part of the output corresponding to the zero padding to be masked.  The final state should also be modified to contain the last unmasked hidden state.  This can be implemented between RNN layers as a separate module in forward-only RNNs, but for bidirectional RNNs this is an issue for the backward direction.  

Because of the aforementioned non-zero output for zero input and state, the "initial" state for a padded sequence can be non-zero and vary according to the padding size.  Alleviating this problem would require masking between time steps for the backward RNN (and therefore might as well be implemented for the forward RNN as well).  

If this isn't possible, would it be possible to have an option to set only some of the gate biases to zero and not update them?  For example, for LSTMs to learn long-distance relations the forget gate bias should be initialized to a large number (1 or 2) but setting the cell gate bias to zero would eliminate the backward problem mentioned above.  

I could implement masking between time steps for the AutogradRNN, but I'm afraid I don't understand the CuDNN backend well enough to modify that, if it's even possible.

**Comment by ngimel** [comment]
Cudnn library supports variable length sequences within a batch (provided that input is sorted by sequence length) but this feature is currently not supported in pytorch Cudnn backend. Batch_first for variable length sequences would also be really awkward.

**Comment by braingineer** [comment]
@shawnjhenry , why batch_size, and not a binary matrix that you use to zero out or do boolean switches over the tensors?   I've usually implemented it as a sentinel value in the input data (0, usually), then forming the binary mask based on that.  For any modifications to the data tensor, you just do a corresponding operation to the mask tensor.

**Comment by shawnjhenry** [comment]
@ngimel That actually sounds like what I'm looking for - any chance that will be supported in the future?  How does CuDNN actually handle the variable length sequences?  I assume because of your comment about batch_first, you pass it a list of size max_seq_length consisting of tensors of size <= batch_size x hidden_size?  

Batch first could still work - you pass a list of size batch_size consisting of of tensors of size seq_length_i x hidden_size, and before passing to the CuDNN backend, sort by seq_length and concatenate at each time step to get it into the first form.  You would need to keep track of the permutation of batches and then do the inverse permutation on the output, but there are cases where you might need to do that for the regular CuDNN input anyway.

@braingineer The lengths are necessary because whether or not to mask a certain row is dependent on the time step.  I suppose you could provide the full mask ahead of time and slice it, but it would require a lot less memory just to pass the lengths and compute the mask on the fly, especially for extremely long sequences or extremely large batch sizes.

**Comment by ngimel** [comment]
Please refer to cudnn manual for information on how it is supported. Generally, cudnn expects an array of input descriptors as an input to forward/backward calls, and those input descriptors specify what batch size is present for this step in time sequence (e.g. you start with a batch size of 64 for step 0, then shorter sequences drop out, and at step, say, 30 you batch size is 42, at 31 batch size is 40 etc - as I've said, the input has to be sorted by sequence length). The input (and output) data is no longer a 3D tensor, it is expected to be stacked 2D tensors with the sizes [[N1, I], [N2,I], [N3,I] ... ] where N_i is number of samples in the batch at i-th step, I is input feature size. N_i is supposed to be sorted in decreasing order. Currently, for batch_first inputs and outputs are just transposed - it will no longer be possible. But the solution you outline should work.

**Comment by braingineer** [comment]
> The lengths are necessary because whether or not to mask a certain row is dependent on the time step.

I'm confused by this statement. Why do sentinel values not do this?  e.g. [this embedding layer in keras](https://github.com/fchollet/keras/blob/master/keras/layers/embeddings.py#L105).  

> especially for extremely long sequences or extremely large batch sizes.

A binary tensor isn't all that big.   Even at 16 bit ints, if you had batch size 128, seq length 100, and feature dim of 300, the mask is 12 mb.  To store, not huge. to apply, usually it's a [switch](https://github.com/fchollet/keras/blob/master/keras/backend/theano_backend.py#L1082) (which is essentially ["where" in numpy](https://docs.scipy.org/doc/numpy/reference/generated/numpy.where.html))

Though, in looking at the code here, it seems it may be harder than I had thought. I'm not finding a function with `where` or `switch` semantics..

@ngimel , is there a function that provides this functionality?  I'll open an issue if not. 

Thanks!

**Comment by jekbradbury** [comment]
Chainer's wrapper for cudnn LSTMs ([L.NStepLSTM](https://github.com/pfnet/chainer/blob/master/chainer/links/connection/n_step_lstm.py) itself wrapping [F.NStepLSTM](https://github.com/pfnet/chainer/blob/master/chainer/functions/connection/n_step_lstm.py)) takes full advantage of cudnn's variable-length sequence support. But the wrapper is complex and fairly slow. Might be worth looking at for inspiration though @colesbury @adamlerer

**Comment by ngimel** [comment]
Or if you guys think cudnn should handle variable length sequences differently, we are open for suggestions. @jekbradbury, @adamlerer.

**Comment by adamlerer** [comment]
Yes, Pytorch should add support for variable length sequences. I think the way CuDNN does it now is fine. I'll have to think about how to efficiently support it on the non-cuDNN side. @soumith you can assign this to me.

**Comment by shawnjhenry** [comment]
@braingineer Using `where` or `switch` won't work in this situation because you might have inputs that are zero but shouldn't be masked, since the mask is also applied to the incoming gradient.  Masking should only happen in specified positions, which, one way or another, requires a list of lengths.

**Comment by shawnjhenry** [comment]
@adamlerer While you're contemplating how to add variable length sequences to the non-CuDNN RNN library, would you consider some other changes?  I've profiled a lot of RNN libraries, and it turns out that most of the calculation time is wasted in kernel launches when operating on the GPU.  Pre-allocating buffer tensors and doing the calculations in-place on slices, as in [this torch library](https://github.com/jcjohnson/torch-rnn) is considerably faster than having to create new tensors at each time step.  It's a bit more involved to write, but as you can see from the benchmarks there, the speed-up and memory savings is significant, especially for one-layer nets.

**Comment by braingineer** [comment]
@shawnjhenry incoming gradient? I thought it was a forward pass masking?  Sorry if seeming slow on this, I think I've been spoiled by theano, and all of the theano-based libraries I've worked with have used the sentinel-based mask.

**Comment by shawnjhenry** [comment]
@braingineer When doing backpropagation through time, at time step `t` the gradient is computed using the gradient of time step `t+1` with respect to the hidden state(s), which are the output(s) of time step `t`.  Therefore, the gradient from `t+1` has to be masked identically to the output of `t`.

Masking using sentinel values seems like a bad idea to me.  In the (admittedly unlikely if you do things carefully but you never know) situation where a coordinate that shouldn't be masked happens to take on the sentinel value, you'll prevent that coordinate from learning.

For instance, if you compute the mask immediately after and embedding lookup and use a sentinel value of zero, if any of the embeddings that shouldn't be masked happen to have zeros in them, those coordinate will never learn.

**Comment by shawnjhenry** [comment]
@braingineer In case you were wondering, sentinel masking *is* possible in PyTorch: you can do:
``` 
mask = torch.ne(tensor, sentinel)
```
which returns a `ByteTensor` of the same size as `tensor` with a 1 in the positions where the comparison is true.

**Comment by adamlerer** [comment]
@shawnjhenry 
> While you're contemplating how to add variable length sequences to the non-CuDNN RNN library, would you consider some other changes? I've profiled a lot of RNN libraries, and it turns out that most of the calculation time is wasted in kernel launches when operating on the GPU. 

Yes, unless you use CuDNN, which combines multiple iterations into a single launch call, you spend most of your time doing kernel launches. You can be clever and try to reduce kernel launches (e.g. @apaszke noticed that you can do `W_ih * x` for the entire sequence in parallel) but these optimizations trade off speed for flexibility. I have been assuming that if you want speed you'll use CuDNN so I haven't yet optimized the non-CuDNN backend much; it's goal was to be simple, flexible, and match the cuDNN RNN for testing. The non-CuDNN backend has similar performance to the Lua-Torch RNN library I compared to, but I'd be curious to compare with torch-rnn to see

> Pre-allocating buffer tensors and doing the calculations in-place on slices, as in this torch library is considerably faster than having to create new tensors at each time step.

Are you sure that's still a bottleneck in PyTorch? That certainly used to be the case in Lua Torch, but the new caching CUDA caching allocator is very fast, and we typically don't see memory allocation affecting performance any more.

**Comment by rjkmurray** [comment]
You when you build with me

**Comment by shawnjhenry** [comment]
@adamlerer Good to know about the caching allocator.  I haven't actually looked to see what bottlenecks exist in PyTorch yet - I was thinking about about Lua Torch.  Before I saw the library I linked I built a similar one and saw up to a 5x speedup over the standard Torch RNN library depending on the input size.  The biggest use of time in the standard library seems to be memory allocation, which is why I thought it might be an issue here, but it sounds like that is not the case.

In case you're interested, I did some testing, and found that indexing is considerably faster than masking, at least for the vanilla tanh rnn.  Indexing also works by passing a `length` tensor as I mentioned above, but instead of using it to create a mask, I create an index at step `i`: 
```
index = torch.squeeze(torch.nonzero(torch.lt(i, length)))
```
and use it to do an index select: 
```
step_input = torch.index_select(input[i], 0, index)
step_hidden = torch.index_select(hidden, 0, index)
```
then after computing:
```
step_hidden = inner(step_input, step_hidden, *weight)
```
I scatter `step_hidden` into a zero tensor:
```
hidden = input.new(hidden.size()).fill_(0)
hidden.scatter_(0, index.unsqueeze(1).expand_as(step_hidden), step_hidden)
```
and after the output is concatenated after last time step, I get the final state using gather:
```
index = lengths.unsqueeze(0).unsqueeze(2).expand(1, output.size(1), output.size(2))
hidden = torch.gather(output, 0, index)
```
all inside the `forward` function in `Recurrent`.  For inputs of size `(100, 50, 1024)` and lengths sampled randomly, this method is, on average, about 25% faster than the equivalent method using a binary mask.  Unfortunately the last step won't work for LSTM since nothing is keeping track of the `c` state, so that needs a bit of a rethink.

**Comment by soumith** [comment]
implemented in https://github.com/pytorch/pytorch/releases/tag/v0.1.10

**Comment by nitishgupta** [comment]
@soumith : Can such a behavior be achieved without having to sort the batch according to lengths of sequences? Also, where can I find more information on how the RNN/LSTM works when the input is a PaddedSequence?

**Comment by MaigoAkisame** [comment]
Just an update for future readers: Now PyTorch has the functions pack_padded_sequence and pad_packed_sequence for RNNs to process sequences with different lengths in a batch.

**Comment by tongjinle123** [comment]
But sometimes we have more than one input. We can't sort both seq1 and seq2 at the same time. Like machine comprehension model, our inputs is: question-passage-label.

**Comment by ozancaglayan** [comment]
Hello,

I see that this was fixed by the introduction of `PackedSequences` and the ability of CUDNN to handle variable-length batches *but* isn't the issue still out there for bidirectional RNNs? Although you make sure that the embedding for the `<pad>` token is 0, due to the existence of biases, starting backward encoding from a sequence of `<pad>` tokens will produce wrong results. I created an MWE for this: 

```python
import warnings
from torch import nn
from torch.nn.utils.rnn import *

warnings.filterwarnings('ignore')


if __name__ == '__main__':
    torch.manual_seed(234)
    torch.cuda.manual_seed(234)

    emb_layer = nn.Embedding(10, 5, padding_idx=0).cuda()
    rnn = nn.GRU(5, 1, bidirectional=True).cuda()

    # Tie forward-backward weights to obtain the same outputs for
    # a symmetrical input
    rnn.weight_ih_l0_reverse = rnn.weight_ih_l0
    rnn.weight_hh_l0_reverse = rnn.weight_hh_l0
    rnn.bias_ih_l0_reverse = rnn.bias_ih_l0
    rnn.bias_hh_l0_reverse = rnn.bias_hh_l0

    def print_states(tag, states):
        msg = ' -> '.join([f'{s.cpu().item():.4f}' for s in states])
        print(f'{tag}: {msg}')

    # disable gradients
    with torch.no_grad():
        # fill padding embeddings with 0
        emb_layer.weight[0].fill_(0)

        print('No padding, sequence is: [2, 4, 2], bw and fw states should be symmetrical')
        batch = pad_sequence([torch.LongTensor([2, 4, 2])]).cuda()
        embs = emb_layer(batch)
        hids = rnn(embs)[0].squeeze_(1)
        print_states('  fw_states', hids[:, :1])
        print_states('  bw_states', hids[:, 1:])
        print()

        print('Explicit padding -> packing and then padding. sequence is: [2, 4, 2, 0, 0]')
        print('Results match, paddings ignored correctly for fw and bw')
        batch = pad_sequence([torch.LongTensor([2, 4, 2, 0, 0])]).cuda()
        embs = emb_layer(batch)
        packed_embs = pack_padded_sequence(
            embs, lengths=batch.ne(0).sum(0),
            enforce_sorted=False)

        hids = pad_packed_sequence(rnn(packed_embs)[0])[0].squeeze_(1)
        print_states('  fw_states', hids[:, :1])
        print_states('  bw_states', hids[:, 1:])
        print()

        print('Same as above, without packing')
        print('fw_states correct, the last two padded computations can be ignored with masking')
        print('bw_states completely wrong as it encodes [0, 0, 2, 4, 2]')
        hids = rnn(embs)[0].squeeze_(1)
        print_states('  fw_states', hids[:, :1])
        print_states('  bw_states', hids[:, 1:])
```

## Expected behavior

Output from the above MWE is as follows. The last experiment shows that the backward states obtained without using non-packed data are wrong. CPU results also replicate the below ones, there is no change in behavior.


```
No padding, sequence is: [2, 4, 2], bw and fw states should be symmetrical
  fw_states: 0.4392 -> 0.1081 -> 0.5033
  bw_states: 0.5033 -> 0.1081 -> 0.4392

Explicit padding -> packing and then padding. sequence is: [2, 4, 2, 0, 0]
Results match, paddings ignored correctly for fw and bw
  fw_states: 0.4392 -> 0.1081 -> 0.5033
  bw_states: 0.5033 -> 0.1081 -> 0.4392

Same as above, without packing
fw_states correct, the last two padded computations can be ignored with masking
bw_states completely wrong as it encodes [0, 0, 2, 4, 2]
  fw_states: 0.4392 -> 0.1081 -> 0.5033 -> 0.5004 -> 0.4987
  bw_states: 0.5343 -> 0.1623 -> 0.5975 -> 0.2766 -> 0.1622
```
